pkg/lightning/restore/check_info.go

```diff
@@ -0,0 +1,501 @@
+					passed = false
+				} else {
+					message = fmt.Sprintf("local disk space is not enough to finish import, source dir has %s, but local available is %s,"+
+						"so, enable disk-quota automatically", units.BytesSize(float64(sourceSize)), units.BytesSize(float64(localAvailable)))
+					passed = true
+					rc.cfg.TikvImporter.DiskQuota = config2.ByteSize(localAvailable - reservedSize)
+				}
+			}
+		}
+	}
+	rc.checkTemplate.Collect(Critical, passed, message)
+	return nil
+}
+
+// CheckpointIsValid checks whether we can start this import with this checkpoint.
+func (rc *Controller) CheckpointIsValid(ctx context.Context, tableInfo *md.MDTableMeta) ([]string, bool, error) {
+	msgs := make([]string, 0)
+	uniqueName := common.UniqueTable(tableInfo.DB, tableInfo.Name)
+	tableCheckPoint, err := rc.checkpointsDB.Get(ctx, uniqueName)
+	if err != nil {
+		// there is no checkpoint
+		log.Debug("no checkpoint detected", zap.String("table", uniqueName))
+		return nil, true, nil
+	}
+	// if checkpoint enable and not missing, we skip the check table empty progress.
+	if tableCheckPoint.Status <= checkpoints.CheckpointStatusMissing {
+		return nil, false, nil
+	}
+
+	var permFromCheckpoint []int
+	var columns []string
+	for _, eng := range tableCheckPoint.Engines {
+		if len(eng.Chunks) > 0 {
+			chunk := eng.Chunks[0]
+			permFromCheckpoint = chunk.ColumnPermutation
+			columns = chunk.Chunk.Columns
+			if filepath.Dir(chunk.FileMeta.Path) != rc.cfg.Mydumper.SourceDir {
+				message := fmt.Sprintf("chunk checkpoints path is not equal to config"+
+					"checkpoint is %s, config source dir is %s", chunk.FileMeta.Path, rc.cfg.Mydumper.SourceDir)
+				msgs = append(msgs, message)
+			}
+		}
+	}
+	if len(columns) == 0 {
+		log.Debug("no valid checkpoint detected", zap.String("table", uniqueName))
+		return nil, false, nil
+	}
+	info := rc.dbInfos[tableInfo.DB].Tables[tableInfo.Name]
+	if info != nil {
+		permFromTiDB, err := parseColumnPermutations(info.Core, columns, nil)
+		if err != nil {
+			msgs = append(msgs, fmt.Sprintf("failed to calculate columns %s, table %s's info has changed,"+
+				"consider remove this checkpoint, and start import again.", err.Error(), uniqueName))
+		}
+		if !reflect.DeepEqual(permFromCheckpoint, permFromTiDB) {
+			msgs = append(msgs, fmt.Sprintf("compare columns perm failed. table %s's info has changed,"+
+				"consider remove this checkpoint, and start import again.", uniqueName))
+		}
+	}
+	return msgs, false, nil
+}
+
+// hasDefault represents col has default value.
+func hasDefault(col *model.ColumnInfo) bool {
+	return col.DefaultIsExpr || col.DefaultValue != nil || !mysql.HasNotNullFlag(col.Flag) ||
+		col.IsGenerated() || mysql.HasAutoIncrementFlag(col.Flag)
+}
+
+// SchemaIsValid checks the import file and cluster schema is match.
+func (rc *Controller) SchemaIsValid(ctx context.Context, tableInfo *md.MDTableMeta) ([]string, error) {
+	msgs := make([]string, 0)
+	info := rc.dbInfos[tableInfo.DB].Tables[tableInfo.Name]
```

Comment:
Will this line panic if rc.dbInfos[tableInfo.DB] is nil when neither db schema file nor target schema in this exists

Reply:
`rc.dbInfos` was generated by `rc.dbmetas` in loadSchemaInfo. so it cannot be nil.